simulator.py

```diff
@@ -0,0 +1,169 @@
+        return result
+
+    def produce_portfolio(self, today, alpha_formula, data):
+        """
+        Find daily portfolio.
+        """
+        portfolio = alpha_formula(today, data)
+        portfolio.normalize()
+        self.__portfolio.append(portfolio)
+
+    def produce_pnl(self, region, cost_function, target_data):
+        """
+        Calculate pnl for alphas.
+        This method will be called by simulation day by day while inputting
+        a new portfolio.
+
+        While calculating pnl,
+        we should take a look at our asset and trading cost.
+            (1) Total asset
+                We assume that our assets are the same everyday (equal to
+                self.initial_asset). To achieve this goal, we open a virtual
+                account to borrow/lend money to the investing account (this
+                account is self.cur_asset[self.base]).
+            (2) Trading cost
+                There are different functions about calculating trading cost
+                in different investing targets.
+                This method should import different modules while targets are
+                different.
+
+                In our template function, we calculate trading cost while
+                selling out the asset.
+                The trading_cost function could be built by alpha writers.
+                The API will adjust in the future, while the fuction needs
+                more data.
+        """
+        today = self.__portfolio[-1]
+        cost = 0
+        trading_value = 0
+        profit = 0
+        for instrument, proportion in today.items():
+            if target_data[instrument][-1].sell is not None:
+                now = self.quantity(
+                    proportion,
+                    proportion / target_data[instrument][-1].sell
+                )
+                trading_value += now.value
+                profit += (
+                    (self.cur_asset[instrument].quantity - now.quantity) *
+                    target_data[instrument][-1].sell)
+                cost += cost_function(
+                    self.cur_asset[instrument],
+                    now,
+                    target_data[instrument][-1]
+                )
+                self.cur_asset[instrument] = now
+        basic_value = 1 - trading_value
+        profit += self.cur_asset[region].value - basic_value
+        self.cur_asset[region] = self.quantity(basic_value, basic_value)
+        self.__pnl.append((today.date, profit - cost, cost))
+
+
+class Simulator(SimulateMethod):
+    """
+    Simulate the alpha.
+    """
+    def __init__(self, settings, data):
+        SimulateMethod.__init__(
+            self,
+            settings.target[1],
+            list(data['target_data'].keys()))
+        self.alpha_formula = None
+        self.data = None
+        self.settings = settings
+        self.target_data = None
+        self.__key = random.randint(0, 2147483647)
```

Comment:
While this is workable here, I would recommend to use a key with higher security level. For example, we can use [UUID](https://docs.python.org/3/library/uuid.html#uuid.uuid4).